### PR TITLE
Fix a bug where we lose an insertion immediately following a deletion…

### PR DESCRIPTION
**Pull request by takutosato**
Suppose we have a deletion immediately followed by an insertion (as happens after realignment when we have an insetion + MNP as shown in the igv screeshot below). At the locus immediately preceding the insertion (i.e. the end position of the deletion and the start position of the insertion), we lose this insertion, because of the following current code behavior: 

a) `EventMap.getOverlappingEvents()` at this locus returns both the insertion and the deletion.
b) In `AssemblyBasedCallerUtils.CreateAlleleMapper()`, we assign the haplotype containing both events to the spanning deletion, then `break` to exit the for loop that traverses overlapping events. As a result the insertion allele is assigned to no haplotypes and gets dropped.

The solution is to modify `EventMap.getOverlappingEvents()` so that in the above scenario, we only return the insertion event. We don't lose the deletion because it's found by `getOverlappingReads` upstream (even when the deletion has length 1). Returning the deletion here is only good for the purpose of obtaining a spanning deletion (*) allele, but if there's also an insertion there the insertion should take priority.

(Top panel: input bam. Bottom panel: HC bamout.)
Screen Shot 2020-07-08 at 3 27 51 PM

### CODE REVIEW

**Comment by takutosato** [comment]
@davidbenjamin could you review this? There isn't much code change.

@jamesemery here it is.

File: src/test/java/org/broadinstitute/hellbender/utils/haplotype/EventMapUnitTest.java [hunk 1]
```diff
@@ -64,6 +66,81 @@ public void testMNPs(final String refBases, final String haplotypeBases, final S
...
+        final Haplotype hap = new Haplotype(haplotypeBases.getBytes(), false, hapStartWrtRef, TextCigarCodec.decode(cigar));
+        final EventMap eventMap = new EventMap(hap, refBases.getBytes(), refLoc, NAME, 1);
+
+        // final List<VariantContext> overlappingEvents = eventMap.getOverlappingEvents(queryLoc);
+        final List<VariantContext> overlappingEvents = eventMap.getOverlappingEventsLegacy(queryLoc);
+
+        final boolean noEventsExpected = expectedAlt == null && expectedRef == null;
+        Assert.assertEquals(overlappingEvents.size(), noEventsExpected ? 0 : 1);
+
+        if (noEventsExpected) {
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/utils/haplotype/EventMapUnitTest.java:133]: Define `eventsExpected = expectedAlt != null || expectedRef != null`, then have
```
if (eventsExpected) {
   Assert. . . 
}
```

**Comment by takutosato** [comment]
Thanks for reviewing @davidbenjamin!

@ldgauthier, this change causes `HaplotypeCallerIntegrationTest.testHaploidNoCall()` to fail—what is going on at this test locus? It looks _very_ messy in IGV...

**Comment by jamesemery** [comment]
@takutosato @davidbenjamin what is the status of this PR? We would like to get these changes in first before we try merging #6634.

**Comment by takutosato** [comment]
@jamesemery I just need to fix one test. Will merge by the end of the week, is that acceptable?

**Comment by takutosato** [comment]
@ldgauthier At the test locus used in `HaplotypeCallerIntegrationTest.testHaploidNoCall()`, <NON_REF> allele/haploid genotype is no longer the most likely genotype after the changes in this PR. So I removed the integration test and added a unit test that checks that when the <NON_REF> allele is present in the genotype, the no_call genotype (e.g. "." or "./.") has the correct ploidy, which was the problem #6563 fixed.

**Pull request merged**